rust/template/d3log/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+    fn deserialize_batch(&self, s: Vec<u8>) -> Result<DDValueBatch, Error>;
+}
+
+pub type Evaluator = Arc<(dyn EvaluatorTrait + Send + Sync)>;
+
+#[derive(Clone)]
+pub enum Batch {
+    Value(DDValueBatch),
+    Rec(RecordBatch),
+}
+
+impl Display for Batch {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "Batch [").unwrap();
+        match self {
+            Batch::Value(b) => b.fmt(f),
+            Batch::Rec(b) => b.fmt(f),
+        }
+        .unwrap();
+        writeln!(f, "\n]\n")
+    }
+}
+
+pub trait Transport {
+    // since most of these errors are async, we're adopting a general
+    // policy for the moment of making all errors async and reported out
+    // of band.
+
+    // should really be type parametric shouldn't it?
+    fn send(&self, b: Batch);
+}
+
+pub type Port = Arc<(dyn Transport + Send + Sync)>;
+
+struct AccumulatePort {
+    eval: Evaluator,
+    b: Arc<Mutex<DDValueBatch>>,
+}
+
+impl Transport for AccumulatePort {
+    fn send(&self, b: Batch) {
+        for (r, f, w) in &DDValueBatch::from(&(*self.eval), b).expect("iterator") {
+            self.b.lock().expect("lock").insert(r, f, w);
+        }
+    }
+}
+
+use std::collections::VecDeque;
+
+struct EvalPort {
+    eval: Evaluator,
+    forwarder: Port,
+    dispatch: Port,
+    queue: Arc<Mutex<VecDeque<Batch>>>,
+}
+
+impl Transport for EvalPort {
+    fn send(&self, b: Batch) {
+        println!(
+            "ep {} {}",
+            self.eval.clone().myself(),
+            RecordBatch::from(self.eval.clone(), b.clone())
+        );
+        self.dispatch.send(b.clone());
+
+        {
+            self.queue.lock().expect("lock").push_back(b.clone());
+        }
+
+        loop {
+            let b = {
+                match self.queue.try_lock() {
+                    Ok(mut x) => match x.pop_front() {
+                        Some(x) => x.clone(),
+                        None => {
+                            return;
+                        }
+                    },
+                    Err(_) => {
+                        return;
+                    }
+                }
+            };
+            let out = async_error!(self.eval.clone(), self.eval.eval(b.clone()));
+            self.dispatch.send(out.clone());
+            self.forwarder.send(out.clone());
+        }
+    }
+}
+
+struct ThreadInstance {
+    rt: Arc<tokio::runtime::Runtime>,
+    eval: Evaluator,
+    evalport: Port,
+    new_evaluator: Arc<dyn Fn(Node, Port) -> Result<(Evaluator, Batch), Error> + Send + Sync>,
+    forwarder: Arc<Forwarder>,
+    broadcast: Arc<Broadcast>,
+    // since all kinds of children will need a copy of the management state,
+    // consider generalizing
+    accumulator: Arc<Mutex<DDValueBatch>>,
+}
+
+// we're just throwing this into the same runtime - do we want/need scheduling isolation?
+// xxx handle deletes
+impl Transport for ThreadInstance {
+    fn send(&self, b: Batch) {
```

Comment:
This is basically `fork`, right?  Let's discuss this face-to-face, I'm struggling to form a big picture here.